src/zenml/integrations/modal/utils.py

```diff
@@ -0,0 +1,642 @@
+    logger.info(
+        f"Serverless application configured with min_containers={effective_min_containers}, max_containers={effective_max_containers}"
+    )
+
+    return execute_step_func, app_name
+
+
+def get_resource_settings_from_deployment(
+    deployment: Any,
+    resource_settings_key: str = "resources",
+) -> ResourceSettings:
+    """Extract resource settings from pipeline deployment.
+
+    Args:
+        deployment: The pipeline deployment.
+        resource_settings_key: Key to look for resource settings.
+
+    Returns:
+        ResourceSettings object with the configuration.
+    """
+    pipeline_resource_settings: Union[Dict[str, Any], Any] = (
+        deployment.pipeline_configuration.settings.get(
+            resource_settings_key, {}
+        )
+    )
+    if pipeline_resource_settings:
+        # Convert to dict if it's a BaseSettings instance
+        if hasattr(pipeline_resource_settings, "model_dump"):
+            pipeline_resource_dict = pipeline_resource_settings.model_dump()
+        else:
+            pipeline_resource_dict = pipeline_resource_settings
+        resource_settings = ResourceSettings.model_validate(
+            pipeline_resource_dict
+        )
+    else:
+        # Fallback to first step's resource settings if no pipeline-level resources
+        if deployment.step_configurations:
+            first_step = list(deployment.step_configurations.values())[0]
+            resource_settings = first_step.config.resource_settings
+        else:
+            resource_settings = ResourceSettings()  # Default empty settings
+
+    return resource_settings
+
+
+def stream_modal_logs_and_wait(
+    function_call: Any,
+    description: str,
+    app_name: str,
+    check_interval: float = 2.0,
+) -> Any:
+    """Stream logs from Modal app using CLI and wait for FunctionCall completion.
+
+    Args:
+        function_call: The Modal FunctionCall object from .spawn()
+        description: Description of the operation for logging.
+        app_name: Name of the Modal app to stream logs from.
+        check_interval: How often to check for completion (seconds).
+
+    Returns:
+        The result of the function execution.
+
+    Raises:
+        Exception: If the Modal function execution fails.
+        KeyboardInterrupt: If the user cancels the execution.
+    """
+    logger.info(f"Starting {description}")
+
+    # Get function call ID for filtering (if available)
+    call_id = None
+    try:
+        call_id = getattr(function_call, "object_id", None)
+        if call_id:
+            logger.debug(f"Function call ID: {call_id}")
+    except Exception:
+        pass
+
+    # Wait a moment for the function to start before beginning log streaming
+    # This helps avoid capturing old logs from previous runs
+    time.sleep(1)
+
+    # Start log streaming in a separate thread
+    log_stream_active = threading.Event()
+    log_stream_active.set()
+    start_time = time.time()
+
+    def stream_logs() -> None:
```

Comment:
Function in function smells a bit wrong and also wondering if we should instead use their Python SDK to stream the logs? https://github.com/modal-labs/modal-client/blob/4177d0b994ac69e01ada7d7a96655c9dcaae570e/modal/cli/utils.py#L24

Possibly something for down the line, though the func-in-func seems off.